awx/main/analytics/core.py

```diff
@@ -82,145 +61,152 @@ def decorate(f):
+        data = {}
+        for func in json_collectors:
             key = func.__awx_analytics_key__
-            path = '{}.json'.format(os.path.join(gather_dir, key))
-            with open(path, 'w', encoding='utf-8') as f:
-                try:
-                    json.dump(func(last_run, collection_type=collection_type, until=until), f)
-                    manifest['{}.json'.format(key)] = func.__awx_analytics_version__
-                except Exception:
-                    logger.exception("Could not generate metric {}.json".format(key))
-                    f.close()
-                    os.remove(f.name)
-        elif func.__awx_analytics_type__ == 'csv':
+            filename = f'{key}.json'
+            try:
+                data[filename] = (func(last_run, collection_type=collection_type, until=until),
+                                  func.__awx_analytics_version__)
+            except Exception:
+                logger.exception("Could not generate metric {}".format(filename))
+        if data:
+            tgzfile = package(dest.parent, data, until)
+            if tgzfile is not None:
+                tarfiles.append(tgzfile)
+
+                if collection_type != 'dry-run':
+                    ship(tgzfile)
+                    with disable_activity_stream():
+                        for filename in data:
+                            last_entries[filename.replace('.json', '')] = until
+                        settings.AUTOMATION_ANALYTICS_LAST_ENTRIES = last_entries
+
+        for func in csv_collectors:
             key = func.__awx_analytics_key__
+            filename = f'{key}.csv'
             try:
-                files = func(last_run, full_path=gather_dir, until=until)
-                if files:
-                    manifest['{}.csv'.format(key)] = func.__awx_analytics_version__
-                if len(files) > num_splits:
-                    num_splits = len(files)
+                slices = [(last_run, until)]
+                if func.__awx_expensive__:
+                    slices = func.__awx_expensive__(key, last_run, until)  # it's ok if this returns a generator
+                for start, end in slices:
+                    files = func(start, full_path=gather_dir, until=end)
+
+                    if not files:
+                        continue
+                    for fpath in files:
+                        tgzfile = package(dest.parent, {filename: (fpath, func.__awx_analytics_version__)}, until)
+                        if tgzfile is not None:
+                            tarfiles.append(tgzfile)
+
+                            if collection_type != 'dry-run':
+                                ship(tgzfile)
+                                with disable_activity_stream():
+                                    last_entries[key] = end
+                                    settings.AUTOMATION_ANALYTICS_LAST_ENTRIES = last_entries
```

Comment:
change to how we did this before:

before: if there was collection error we've skipped the 4h interval (and there is issue about sending report about this)
current: we'll get stuck on the error (so we'll receive no more events), since we're not moving the id marker?

Tbh. not sure what is better, feels like the way it was before is slightly better, although if we're receiving data with holes in it periodically, it's unusable. But at least the recovery doesn't require a DB change? @ryanpetrello thoughts?

Reply:
My understanding of the issue was that this behavior was what we are aiming for.  cc @wenottingham

Reply:
I'd prefer to not update the ID if we don't have confirmed submission of the data.

Reply:
@jbradberry @wenottingham ok, that sounds good, just stating the side effect:

- in a case of error we won't be sending any more events (e.g. like we've seen, when we can't read some unicode chars), until the customer updates their Tower with a fix

And I see we're doing this just for events? The rest is using the common last_run. Should we align all collections like this?

---------

Then with sending a report of failures/successes, this is probably good.